### PR TITLE
KEP-814 Add client key used in the database request to the crud api

### PR DESCRIPTION
**Pull request by ebruck**

### CODE REVIEW

File: crud/crud.cpp [hunk 1]
```diff
@@ -107,13 +95,24 @@ crud::send_response(const database_msg& request, const bzn::storage_base::result
...
 
-    if (this->storage->has(PERMISSION_UUID, request.header().db_uuid()))
+    std::lock_guard<std::shared_mutex> lock(this->lock); // lock for write access
+
+    auto [db_exists, perms] = this->get_database_permissions(request.header().db_uuid());
+
+    if (db_exists)
     {
-        result = this->storage->create(request.header().db_uuid(), request.create().key(), request.create().value());
+        if (!this->is_caller_a_writer(caller_id, perms))
```
Comment (rnistuk) [inline crud/crud.cpp]: ok, does the user have write access..

File: crud/crud.cpp [hunk 2]
```diff
@@ -107,13 +95,24 @@ crud::send_response(const database_msg& request, const bzn::storage_base::result
...
+        {
+            result = bzn::storage_result::access_denied;
+        }
+        else
+        {
+            result = this->storage->create(request.header().db_uuid(), request.create().key(), request.create().value());
```
Comment (rnistuk) [inline crud/crud.cpp]: yup, write

File: crud/crud.cpp [hunk 3]
```diff
@@ -128,11 +127,13 @@ crud::handle_create(const database_msg& request, std::shared_ptr<bzn::session_ba
 
 
 void
-crud::handle_read(const database_msg& request, std::shared_ptr<bzn::session_base> session)
+crud::handle_read(const bzn::caller_id_t& /*caller_id*/, const database_msg& request, std::shared_ptr<bzn::session_base> session)
 {
```
Comment (rnistuk) [inline crud/crud.cpp]: I take it that we are not doing read permissions, at this time... So anybody can read.

File: crud/crud.cpp [hunk 4]
```diff
@@ -153,9 +154,25 @@ crud::handle_read(const database_msg& request, std::shared_ptr<bzn::session_base
 
 
 void
-crud::handle_update(const database_msg& request, std::shared_ptr<bzn::session_base> session)
+crud::handle_update(const bzn::caller_id_t& caller_id, const database_msg& request, std::shared_ptr<bzn::session_base> session)
```
Comment (rnistuk) [inline crud/crud.cpp]: cool, we need permission to update...

File: crud/crud.cpp [hunk 5]
```diff
@@ -153,9 +154,25 @@ crud::handle_read(const database_msg& request, std::shared_ptr<bzn::session_base
...
 {
-    auto result = this->storage->update(request.header().db_uuid(), request.update().key(), request.update().value());
+    bzn::storage_result result{bzn::storage_result::db_not_found};
+
+    std::lock_guard<std::shared_mutex> lock(this->lock); // lock for write access
+
+    const auto [db_exists, perms] = this->get_database_permissions(request.header().db_uuid());
```
Comment (rnistuk) [inline crud/crud.cpp]: I like this method of return multiple values.. I haven't used it yet...

File: crud/crud.cpp [hunk 6]
```diff
@@ -169,9 +186,25 @@ crud::handle_update(const database_msg& request, std::shared_ptr<bzn::session_ba
 
 
 void
-crud::handle_delete(const database_msg& request, std::shared_ptr<bzn::session_base> session)
+crud::handle_delete(const bzn::caller_id_t& caller_id, const database_msg& request, std::shared_ptr<bzn::session_base> session)
```
Comment (rnistuk) [inline crud/crud.cpp]: and delete needs the correct permission...

File: crud/crud.cpp [hunk 7]
```diff
@@ -185,13 +218,15 @@ crud::handle_delete(const database_msg& request, std::shared_ptr<bzn::session_ba
 
 
 void
-crud::handle_has(const database_msg& request, std::shared_ptr<bzn::session_base> session)
+crud::handle_has(const bzn::caller_id_t& /*caller_id*/, const database_msg& request, std::shared_ptr<bzn::session_base> session)
```
Comment (rnistuk) [inline crud/crud.cpp]: has needs no permissions...

File: crud/crud.cpp [hunk 8]
```diff
@@ -202,10 +237,12 @@ crud::handle_has(const database_msg& request, std::shared_ptr<bzn::session_base>
 
 
 void
-crud::handle_keys(const database_msg& request, std::shared_ptr<bzn::session_base> session)
+crud::handle_keys(const bzn::caller_id_t& /*caller_id*/, const database_msg& request, std::shared_ptr<bzn::session_base> session)
 {
     if (session)
```
Comment (rnistuk) [inline crud/crud.cpp]: ditto...

File: crud/crud.cpp [hunk 9]
```diff
@@ -69,32 +73,16 @@ crud::handle_request(const database_msg& request, const std::shared_ptr<bzn::ses
 
 
 void
-crud::send_response(const database_msg& request, const bzn::storage_base::result result,
+crud::send_response(const database_msg& request, const bzn::storage_result result,
     database_response&& response, std::shared_ptr<bzn::session_base>& session)
 {
     *response.mutable_header() = request.header();
```
Comment (rnistuk) [inline crud/crud.cpp]: I've stepped through the code to see what's going on...

File: crud/crud.cpp [hunk 10]
```diff
@@ -226,18 +263,20 @@ crud::handle_keys(const database_msg& request, std::shared_ptr<bzn::session_base
 
 
 void
-crud::handle_size(const database_msg& request, std::shared_ptr<bzn::session_base> session)
+crud::handle_size(const bzn::caller_id_t& /*caller_id*/, const database_msg& request, std::shared_ptr<bzn::session_base> session)
 {
```
Comment (rnistuk) [inline crud/crud.cpp]: ditto

File: crud/crud.cpp [hunk 11]
```diff
@@ -247,7 +286,7 @@ crud::handle_size(const database_msg& request, std::shared_ptr<bzn::session_base
 
 
 void
-crud::handle_subscribe(const database_msg& request, std::shared_ptr<bzn::session_base> session)
+crud::handle_subscribe(const bzn::caller_id_t& /*caller_id*/, const database_msg& request, std::shared_ptr<bzn::session_base> session)
 {
```
Comment (rnistuk) [inline crud/crud.cpp]: subscribe no need for permissions, yet..

File: crud/crud.cpp [hunk 12]
```diff
@@ -311,19 +352,26 @@ crud::handle_create_db(const database_msg& request, std::shared_ptr<bzn::session
 
 
 void
-crud::handle_delete_db(const database_msg& request, std::shared_ptr<bzn::session_base> session)
+crud::handle_delete_db(const bzn::caller_id_t& caller_id, const database_msg& request, std::shared_ptr<bzn::session_base> session)
 {
-    storage_base::result result;
```
Comment (rnistuk) [inline crud/crud.cpp]: needs permissions to delete...

File: crud/crud.cpp (hunk 3 above)
Comment (ebruck) [reply crud/crud.cpp]: yes for mvp

File: crud/crud.cpp (hunk 4 above)
Comment (ebruck) [reply crud/crud.cpp]: it does check for writer

File: crud/crud.cpp (hunk 7 above)
Comment (ebruck) [reply crud/crud.cpp]: same as a read

File: crud/crud.cpp (hunk 8 above)
Comment (ebruck) [reply crud/crud.cpp]: same as a read

File: crud/crud.cpp [hunk 13]
```diff
@@ -353,20 +403,91 @@ crud::handle_has_db(const database_msg& request, std::shared_ptr<bzn::session_ba
...
+{
+    for(const auto& writer_id : json[WRITERS_KEY])
+    {
+        if (writer_id == caller_id)
+        {
+            return true;
+        }
+    }
+
+    return this->is_caller_owner(caller_id, json);
```
Comment (rnistuk) [inline crud/crud.cpp]: ah, nice! If the caller id is not in the writers list, check if the caller id is the owner!

**Pull request merged**